### PR TITLE
New bootstrap 0424

### PR DESCRIPTION
**Pull request by alivedise**

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d10cedb803266eac0b3d341d39d979b571ff298d

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=00532bdbfbe29136352cd5b6ac208f4c3bacecf9

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3d3d77833975ebde4c02840c6f8d4511cd9ff83a

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8229e908f3ed4ffc2c570c1232a689a8f8f1530c

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=88a51b86358fad3ca2579922234323da2ecf99d8

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=de13b316c5b3983d37f069ecddeea59d4da1d1e2

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2dd40eb3ebde683da615db82141de995117132c5

File: apps/system/js/bootstrap.js [hunk 1]
```diff
@@ -1,247 +1,23 @@
-/* -*- Mode: js; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- /
-/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */
+/* global App */
```
Comment (timdream) [inline apps/system/js/bootstrap.js]: How about a similar comment here?

https://github.com/mozilla-b2g/gaia/blob/master/apps/keyboard/js/keyboard/bootstrap.js#L5-L7

File: apps/system/js/app.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      return LazyLoader.load(this.FILES).then(() => {
+        return this.bootstrap();
+      });
+    },
+    bootstrap: function() {
+      if (this._booted) {
+        return Promise.resolve();
+      }
+      this._booted = true;
+      console.log('1111');
```
Comment (timdream) [inline apps/system/js/app.js]: console

File: apps/system/js/app.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      window.settingsCore = BaseModule.instantiate('SettingsCore');
+      window.settingsCore.start();
+      window.launcher = BaseModule.instantiate('Launcher');
+      return Promise.all([
+        applications.waitForReady(), // There is too many operations
```
Comment (timdream) [inline apps/system/js/app.js]: Let's file a bug and get this fixed... I know it blocks on a mozChromeEvent :-/

File: apps/system/js/core.js [hunk 4]
```diff
@@ -51,16 +52,99 @@
...
+          'FindmydeviceLauncher',
+          'FxAccountsManager',
+          'FxAccountsUI',
+          'NetworkActivity',
+          'CrashReporter',
+          'Screenshot',
+          'SoundManager',
+          'CustomDialogService',
+          'CarrierInfoNotifier'
+          // XXX: We should move this into mobileConnectionCore,
```
Comment (timdream) [inline apps/system/js/core.js]: Not entire should what `this` is referring to... L109?

File: apps/system/js/core.js [hunk 5]
```diff
@@ -51,16 +52,99 @@
...
+      // Gecko that we're ready for certain system services to send us
+      // messages (e.g. the radio).
+      // Note that shell.js starts listen for the mozContentEvent event at
+      // mozbrowserloadstart, which sometimes does not happen till
+      // window.onload.
+      this.publish('mozContentEvent', {
+        type: 'system-message-listener-ready'
+      }, true);
+
+      return new Promise((resolve, reject) => {
```
Comment (timdream) [inline apps/system/js/core.js]: This is not the right way to chain promise.

Do you intend for the main chain only resolves when `this.loadWhenIdle()` resolves?
This should be

```javascript
return this.loadWhenIdle([ ... ])
  .then(() => {
    return Promise.all([ ... ]); //L114
  })
  .then(() => {
    // L126, although I am not sure if `developerHud.start();` will return a promise in which we should wait.
  });
```

File: apps/system/js/launcher.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+          homescreenManifestURL, true).then(() => {
+            this.service.request('LogoManager:animatePoweronLogo');
+            this.scheduler.release();
+          })
+      ]).catch((err) => {
+        console.error(err);
+      });
+    },
+    launchLockscreenThenHomescreen: function(homescreenManifestURL) {
+      Promise.all([
```
Comment (timdream) [inline apps/system/js/launcher.js]: Please explain why you would like to start a new chain in a comment above this line.

File: apps/system/js/launcher.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+          this.service.request('HomescreenLauncher:launch',
+            homescreenManifestURL);
+          this.scheduler.release();
+        })
+      ]).catch((err) => {
+        console.error(err);
+      });
+    },
+    launchFtuThenHomescreen: function(ftuManifestURL, homescreenManifestURL) {
+      Promise.all([
```
Comment (timdream) [inline apps/system/js/launcher.js]: Please explain why you would like to start a new chain in a comment above this line.

File: apps/system/js/launcher.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        // We still need to tell FtuLauncher to skip to process some tasks.
+        this.service.request('FtuLauncher:skip'),
+        this.service.request('WallpaperManager:initializeWallpaper',
+          this.wallpaper, this.wallpaperValid).then(() => {
+          return this.service.request('LockScreenLauncher:launch');
+        }).then(() => {
+          this.service.request('LogoManager:animatePoweronLogo');
+          this.service.request('HomescreenLauncher:launch',
+            homescreenManifestURL);
+          this.scheduler.release();
```
Comment (timdream) [inline apps/system/js/launcher.js]: Do you want to chain L108-L111 into the chain? If so this should be wrapped into yet another `Promise.all()`

File: apps/system/js/launcher.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      });
+    },
+    launchFtuThenHomescreen: function(ftuManifestURL, homescreenManifestURL) {
+      Promise.all([
+        this.service.request('FtuLauncher:launch',
+          ftuManifestURL).then(() => {
+            this.service.request('HomescreenLauncher:launch',
+              homescreenManifestURL);
+            this.service.request('LogoManager:animatePoweronLogo');
+            this.scheduler.release();
```
Comment (timdream) [inline apps/system/js/launcher.js]: Do you want to chain L121-L124 into the chain? If so this should be wrapped into yet another `Promise.all()`

File: apps/system/js/launcher.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+          }).catch(function(err) {
+            console.error(err);
+          }),
+        this.service.request('WallpaperManager:initializeWallpaper',
+          this.wallpaper, this.wallpaperValid),
+        this.service.request('stepReady', 'done').then(() => {
+          this.service.request('LockScreenLauncher:standby');
```
Comment (timdream) [inline apps/system/js/launcher.js]: chain by simply `return`?

File: apps/system/js/scheduler.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+          });
+        } else {
+          this.debug('not blocking, execute right away');
+          request();
+          resolve();
+        }
+      });
+    }
+  });
+}());
```
Comment (timdream) [inline apps/system/js/scheduler.js]: whitespace at EOL

File: apps/system/js/scheduler.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    execute: function() {
+      var length = this._queue.length;
+      this.debug('process operations in the schedule pool: ' + length);
+      this._queue.forEach(function(operation) {
+        operation.request();
+        operation.resolve();
+      });
+      this._queue = [];
+    },
+    schedule: function(request) {
```
Comment (timdream) [inline apps/system/js/scheduler.js]: `request` -> `callback`

File: apps/system/js/scheduler.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+      return new Promise((resolve) => {
+        if (this._busyState) {
+          this.debug('busy loading, put request in the pool.');
+          this._queue.push({
+            resolve: resolve,
+            request: request
+          });
```
Comment (timdream) [inline apps/system/js/scheduler.js]: This is essentially the Deferred object; you can decide whether or not to switch.

Comment (timdream) [inline apps/system/js/scheduler.js]: On what, it's not ... sorry about that.

File: apps/system/js/scheduler.js (hunk 12 above)
Comment (timdream) [inline apps/system/js/scheduler.js]: It's not easy to figure out what `request` is from naming of this variable.

File: apps/system/js/base_module.js [hunk 14]
```diff
@@ -111,6 +103,15 @@
...
 
   var SubmoduleMixin = {
+    loadWhenIdle: function(modules) {
+      return new Promise((resolve, reject) => {
+        this.service.request('schedule', () => {
+          this.constructor.SUB_MODULES =
+            this.constructor.SUB_MODULES.concat(modules);
+          this._startSubModules().then(resolve).catch(reject);
+        });
+      });
```
Comment (timdream) [inline apps/system/js/base_module.js]: Maybe this is equivalent to:

```javascript

return this.service.request('schedule', () => {
  this.constructor.SUB_MODULES =
    this.constructor.SUB_MODULES.concat(modules);
  return this._startSubModules();
});

```
?

The fact that `request` and `schedule` are both promises prevent me to understanding this clearly.

**Comment by timdream** [comment]
I dug into `LazyLoader` and this is my finding:

1. It *never* rejects unless the code changes locally.
2. JS error in the loaded script will not be trapped in the LazyLoader promise as well -- no `catch()` needed.

So unless we want to wait for the LazyLoader to load, we don't need to chain it nor `catch()` it. That's less work to do here than previously expected.

File: apps/system/js/base_module.js [hunk 15]
```diff
@@ -118,37 +119,60 @@
...
-        this.constructor.SUB_MODULES.forEach(function(module) {
-          var moduleName = BaseModule.lowerCapital(module);
-          var parent = this.constructor.SUB_MODULE_PARENT || this;
-          if (!parent[moduleName]) {
-            this._initialSubModule(moduleName, module);
-          }
-        }, this);
-      }.bind(this));
+        unloaded.concat());
+      return new Promise((resolve, reject) => {
```
Comment (timdream) [inline apps/system/js/base_module.js]: Remove this `new Promise()` block and simply return the promise from `Promise.all()` at L156.

File: apps/system/js/bluetooth_core.js [hunk 16]
```diff
@@ -16,20 +16,22 @@
...
+            window.Bluetooth = Bluetooth1;
+            window.Bluetooth.start().then(resolve);
+          });
+        } else { // APIv2
+          LazyLoader.load(['js/bluetooth_v2.js'], function() {
+            window.Bluetooth = new Bluetooth2();
+            window.Bluetooth.start().then(resolve);
+          });
+        }
+      });
```
Comment (timdream) [inline apps/system/js/bluetooth_core.js]: Remove the `new Promise()` wrap:

```javascript

      var p;
      // Init Bluetooth module by API version.
      if (typeof(window.navigator.mozBluetooth.onattributechanged) ===
        'undefined') { // APIv1
        p = LazyLoader.load(['js/bluetooth.js'].then(function() {
          window.Bluetooth = Bluetooth1;
          return window.Bluetooth.start();
        });
      } else { // APIv2
        p = LazyLoader.load(['js/bluetooth_v2.js']).then(function() {
          window.Bluetooth = new Bluetooth2();
          return window.Bluetooth.start();
        });
      }

```

File: apps/system/js/accessibility.js [hunk 17]
```diff
@@ -346,9 +346,14 @@
       this.cancelHints();
       if (aDetails.eventType === 'quicknav-menu') {
         if (!this.quicknav) {
-          this.quicknav = new AccessibilityQuicknavMenu();
+          LazyLoader.load(['js/accessibility_quick_nav_menu.js'])
+            .then(function() {
+              this.quicknav = new AccessibilityQuicknavMenu();
+              this.quicknav.show();
+            }.bind(this));
```
Comment (timdream) [inline apps/system/js/accessibility.js]: `.catch()` to show error from L350-351

File: apps/system/js/app_install_manager.js [hunk 18]
```diff
@@ -28,7 +26,9 @@ var AppInstallManager = {
   },
 
   init: function ai_init() {
-    this.systemBanner = new SystemBanner();
+    LazyLoader.load(['js/system_banner.js']).then(() => {
+      this.systemBanner = new SystemBanner();
+    });
```
Comment (timdream) [inline apps/system/js/app_install_manager.js]: `.catch()` to get errors from L30.

File: apps/system/js/app_usage_metrics.js [hunk 19]
```diff
@@ -724,44 +725,46 @@
...
-        onerror: retry,
-        onabort: retry,
-        ontimeout: retry
+        request.send({
+          timeout: AUM.REPORT_TIMEOUT,
+          onload: onload,
+          onerror: retry,
+          onabort: retry,
+          ontimeout: retry
+        });
```
Comment (timdream) [inline apps/system/js/app_usage_metrics.js]: `.catch()` to get errors from this block.

File: apps/system/js/app_window_manager.js [hunk 20]
```diff
@@ -483,210 +460,215 @@
...
+    '_handle_shrinking-start': function(evt) {
+      if (this.shrinkingUI && this.shrinkingUI.isActive()) {
+        return;
+      }
+      if (this.service.query('getTopMostUI') !== this) {
+        return;
+      }
+      if (!window.ShrinkingUI) {
+        LazyLoader.load(['shared/js/shrinking_ui.js']).then(
+          this.launchShrinkingUI.call(this)).catch(function(err) {
```
Comment (timdream) [inline apps/system/js/app_window_manager.js]: `call` -> `bind`?

File: apps/system/js/attention_window_manager.js [hunk 21]
```diff
@@ -115,6 +116,14 @@
       window.addEventListener('rocketbar-overlayopened', this);
       window.addEventListener('languagechange', this);
       Service.request('registerHierarchy', this);
+      if (navigator.mozTelephony) {
+        BaseModule.lazyLoad(['DialerAgent']).then(function() {
+          // DialerAgent will create callscreen window
+          // so it is 'CallscreenWindowLauncher' exactly.
+          var dialerAgent = new DialerAgent();
+          dialerAgent.start();
+        });
```
Comment (timdream) [inline apps/system/js/attention_window_manager.js]: `.catch` to get errors from L123-124.

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=de4782a11b587c822f08e1a9cb181d4fe9351102

File: apps/system/js/bluetooth_transfer.js [hunk 22]
```diff
@@ -219,15 +216,13 @@ var BluetoothTransfer = {
           }
         },
         cancel,
-        confirm,
-        screen
-      )
-      .setAttribute('data-z-index-level', 'system-dialog');
+        confirm
+      );
     });
```
Comment (timdream) [inline apps/system/js/bluetooth_transfer.js]: `.catch()` for L208-220

File: apps/system/js/bluetooth_transfer.js [hunk 23]
```diff
@@ -602,71 +592,72 @@ var BluetoothTransfer = {
...
+            return;
+          }
+        };
+        a.onsuccess = (e) => {
+          var msg = 'open activity onsuccess';
+          this.debug(msg);
+        };
       };
-    };
+    });
```
Comment (timdream) [inline apps/system/js/bluetooth_transfer.js]: `catch()` for L595-647

File: apps/system/js/browser_context_menu.js [hunk 24]
```diff
@@ -14,7 +14,9 @@
    */
   var BrowserContextMenu = window.BrowserContextMenu = function(app) {
     this.app = app;
-    this._view = new ContextMenuView(app);
+    LazyLoader.load(['js/context_menu_view.js']).then(() => {
+      this._view = new ContextMenuView(app);
+    });
```
Comment (timdream) [inline apps/system/js/browser_context_menu.js]: `.catch()` for L18

File: apps/system/js/call_forwarding.js [hunk 25]
```diff
@@ -203,15 +203,21 @@
...
+          SettingsHelper('ril.cf.enabled',
+            this._defaultCallForwardingIconStates);
+
+        this._addEventHandlers();
+
+        // Initialize the icon states
+        this._slots.forEach(function(slot) {
+          this._initCallForwardingState(slot);
+        }, this);
+      });
```
Comment (timdream) [inline apps/system/js/call_forwarding.js]: Before we could switch to BaseModule, `catch()` for L210-219.

File: apps/system/js/captive_portal.js [hunk 26]
```diff
@@ -31,13 +32,15 @@ var CaptivePortal = {
...
-      this.entrySheet = new EntrySheet(document.getElementById('screen'),
-                                      url,
-                                      new BrowserFrame({url: url}));
-      this.entrySheet.open();
+      LazyLoader.load(['js/entry_sheet.js']).then(function() {
+        this.entrySheet = new EntrySheet(document.getElementById('screen'),
+                                        url,
+                                        new BrowserFrame({url: url}));
+        this.entrySheet.open();
+      }.bind(this));
```
Comment (timdream) [inline apps/system/js/captive_portal.js]: `.catch` for L39-L43.

File: apps/system/js/core.js [hunk 27]
```diff
@@ -75,9 +159,9 @@
           }
           if (!this[moduleName]) {
             reject();
+            return;
           }
-          this[moduleName].start && this[moduleName].start();
-          resolve();
+          this[moduleName].start && this[moduleName].start().then(resolve);
```
Comment (timdream) [inline apps/system/js/core.js]:
```javascript

      return BaseModule.lazyLoad([handler]).then(function() {
        var moduleName = BaseModule.lowerCapital(handler);
        if (window[handler] && typeof(window[handler]) === 'function') {
          this[moduleName] = new window[handler](navigator[api], this);
        } else {
          this[moduleName] =
            BaseModule.instantiate(handler, navigator[api], this);
        }
        if (!this[moduleName]) {
          throw new Error('Core: module not found.');
          return;
        }
        if (this[moduleName].start) {
          return this[moduleName].start();
        }
      }.bind(this));

```

File: apps/system/js/crash_reporter.js [hunk 28]
```diff
@@ -99,9 +99,10 @@ var CrashReporter = (function() {
         }
       };
     }
-
-    var systemBanner = new SystemBanner();
-    systemBanner.show(message, button);
+    LazyLoader.load(['js/system_banner.js']).then(() => {
+      var systemBanner = new SystemBanner();
+      systemBanner.show(message, button);
+    });
```
Comment (timdream) [inline apps/system/js/crash_reporter.js]: `.catch` for L103-104

File: apps/system/js/custom_dialog_service.js [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+    'hideCustomDialog'
+  ];
+  BaseModule.create(CustomDialogService, {
+    name: 'CustomDialogService',
+    showCustomDialog: function(title, msg, cancel, confirm) {
+      return LazyLoader.load(['shared/js/custom_dialog.js']).then(function() {
+        CustomDialog.show(title, msg, cancel, confirm,
+          document.getElementById('screen'))
+                    .setAttribute('data-z-index-level', 'system-dialog');
+      }.bind(this));
```
Comment (timdream) [inline apps/system/js/custom_dialog_service.js]: `.catch` for L19-22.

I am not sure all caller of `showCustomDialog` will catch the error.

File: apps/system/js/device_storage_watcher.js [hunk 30]
```diff
@@ -0,0 +1,169 @@
...
+          value: space.size,
+          unit: space.unit
+        });
+      } else {
+        notification = msg + this._('unknown-free-space');
+      }
+      LazyLoader.load(['js/system_banner.js']).then(() => {
+        var systemBanner = new SystemBanner();
+        systemBanner.show(notification);
+      });
```
Comment (timdream) [inline apps/system/js/device_storage_watcher.js]: `catch()`

File: apps/system/js/launcher.js (hunk 8 above)
Comment (timdream) [inline apps/system/js/launcher.js]: and return the promise of course.

**Pull request closed without merging**